### ISSUE TITLE
[no-floating-promises] Should not report on async IIFEs

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#647]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

```TS
(async () => {
	await something();
})();
```

**Expected Result**

I expected it not to report as it's an intentional action. It doesn't catch a mistake.

If you're not willing to add this exception by default, at least add an option to ignore async IIFEs.

**Actual Result**

It reports as a floating promise, which is technically correct, but not wanted behavior, at least from my perspective.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.11.0` |
| `@typescript-eslint/parser`        | `1.11.0` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.16.0` |
| `npm`                              | `6.9.0` |

**Comment by Jessidhia** [comment]
Using the `void` keyword makes this explicit, and the promise does not count as floating: `void (async () => { /* ... */ })()`

**Comment by bradzacher** [comment]
I would be okay with this living as an option.

It is potentially a reportable error because you might be using an IIFE to enclose some logic + variables with the intention of awaiting the response. I've seen this pattern before because it saves you passing arguments in, and it avoids the locally scoped variables from leaking into the wider function. 

Whilst _you_ have intentionally done this pattern, it doesn't mean everyone will intentionally float a promise there.

**Comment by sindresorhus** [comment]
You're right. An option would be best.

**Comment by glen-84** [comment]
@Jessidhia,
 
> Using the `void` operator makes this explicit, and the promise does not count as floating: `void (async () => { /* ... */ })()`

While this is true in TSLint ([at least for now](https://github.com/palantir/tslint/issues/4653)), it doesn't appear to work in ESLint. Would this be considered a bug, or an enhancement request?
